Integrate Alonzo into Cardano, as a Minimum Viable Product

nfrisby opened the pull request:
Fixes #2871. That is CAD-1720.

Add AlonzoEra to CardanoEras. See the commit messages for more details.

There is some added tech debt here, but only in the tests and an internal debugging tool. We're accepting that for now to avoid blocking the downstream Node teams. See Issue #3139.

nfrisby commented:
This PR currently includes PR #3131 -- once that's mergedt his one should be rebased.

For now, see this PR's diff here https://github.com/input-output-hk/ouroboros-network/compare/nfrisby/alonzo-preparation-2...nfrisby/alonzo-integration

nfrisby commented:
Oh boy. I don't know what that Hydra error is about. I'll return to that tomorrow.

Jasagredo commented on ouroboros-consensus-cardano-test/test/Test/ThreadNet/MaryAlonzo.hs:
```diff
@@ -0,0 +1,386 @@
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
```
This one doesn't seem to be needed from what I can see on this file.

nfrisby replied:
Thanks; removed.

Jasagredo commented on ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:
```diff
@@ -261,26 +263,35 @@ instance CardanoHardForkConstraints c => CanHardFork (CardanoEras c) where
         -- Allegra <-> Mary, ...
-      $ TCons (SelectSameProtocol :* Nil)
+      $ TCons (SelectSameProtocol :* SelectSameProtocol :* Nil)
         -- Mary <-> ...
```
I would maintain the same comment format
```suggestion
        -- Mary <-> Alonzo, ...
```

nfrisby replied:
Fixed!

Jasagredo commented on ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:
```diff
@@ -521,3 +543,47 @@ translateValidatedTxAllegraToMaryWrapper ::
+
+translateTxMaryToAlonzoWrapper ::
+     PraosCrypto c
+  => Alonzo.AlonzoGenesis
```
I'm confused by the fact that `getAlonzoTranslationContext` returns sth that we name as `ctxt` but it is specifically an `Alonzo.AlonzoGenesis`. From what I can see, `shelleyLedgerTranslationContext :: !(Core.TranslationContext era)`, which is associated on their side as:
```
type instance TranslationContext (AlonzoEra c) = AlonzoGenesis
```

I think it would be clearer to make this receive an argument of type `TranslationContext (AlonzoEra c)` and in an abstract way give it to `SL.translateEra` as we "don't care" what that type is from our side. We just use it to give it to this function, and it is a `TranslationContext` and we name it `ctxt`.

nfrisby replied:
I wrote `AlonzoGenesis` instead of `TranslationContext (AlonzoEra c)` because I _generally_ avoid writing type family applications that can be immediately simplified. (And not just avoiding beta reduxes; I mean the result is _simpler_.)

But I did name the function and the variables "context", since that is _why_ it's being passed.

I admit that's a mismash of choices I've made in our code here.

Thanks for the note, I'll consider it with Pawel later day.

nfrisby replied:
Pawel and I decided to leave it as is. That's both of our intuitions. We discussed a few reasons, which we'll talk to you about. But mainly: there's a slight modularity gain to doing what you said, but at least in this specific instance, we're favoring the transparency of using the Alonzo-specific types in these Alonzo-specific contexts.

Jasagredo commented on ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:
```diff
@@ -521,3 +543,47 @@ translateValidatedTxAllegraToMaryWrapper ::
+translateValidatedTxMaryToAlonzoWrapper ::
+     forall c.
+     PraosCrypto c
+  => Alonzo.AlonzoGenesis
```
Same as above.

nfrisby replied:
Replied at https://github.com/input-output-hk/ouroboros-network/pull/3138#discussion_r632438113

Jasagredo commented on ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:
```diff
@@ -521,3 +543,47 @@ translateValidatedTxAllegraToMaryWrapper ::
+
+getAlonzoTranslationContext ::
+     WrapLedgerConfig (ShelleyBlock (AlonzoEra c))
+  -> Alonzo.AlonzoGenesis
```
Same as below.

nfrisby replied:
Replied at https://github.com/input-output-hk/ouroboros-network/pull/3138#discussion_r632438113

Jasagredo commented on ouroboros-consensus-cardano/tools/db-analyser/Block/Cardano.hs:
```diff
@@ -81,14 +85,20 @@ parseCardanoArgs :: Parser CardanoBlockArgs
      Byron.Genesis.Config
   -> Maybe PBftSignatureThreshold
   -> ShelleyGenesis StandardShelley
+  -> SL.AlonzoGenesis
```
Same as in `CanHardFork.hs`. I would call this `Core.TranslationContext (AlonzoEra c)` or (not sure if this would blow things up) say `Core.TranslationContext (AlonzoEra c) ~ SL.AlonzoGenesis` to make explicit that we are relying on that equality to hold.

nfrisby replied:
Replied at https://github.com/input-output-hk/ouroboros-network/pull/3138#discussion_r632438113

Jasagredo reviewed (commented):
Looks like this is going the right way, just left some minor comments related to naming mostly

EncodePanda commented on ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Eras.hs:
```diff
@@ -104,6 +115,7 @@ class ( SL.ShelleyBasedEra era
       , HasField "_nOpt" (Core.PParams era) Natural
       , HasField "_rho" (Core.PParams era) SL.UnitInterval
       , HasField "_tau" (Core.PParams era) SL.UnitInterval
+
```
new line?

nfrisby replied:
Ah yes. I was indecisive here. This was stale. Thanks.

Fixed!

EncodePanda commented on ouroboros-network/src/Ouroboros/Network/NodeToNode/Version.hs:
```diff
@@ -52,6 +52,7 @@ data NodeToNodeVersion
     -- ^ Changes:
     --
     -- * new 'KeepAlive' codec
+    -- * Enable @CardanoNodeToNodeVersion5@, i.e., Alonzo
```
I'm wondering why this is not `NodeToNodeV_8`. Was it because the `NodeToNodeV_7` was not yet released?

nfrisby replied:
Yes; double-checking the versions was on my list for us to TODO today.

nfrisby replied:
Node 1.27.0 uses 9b279c7548ee549e1ed755cd1acb69b6e69d0c7b, which already includes `NodeToNodeV_7`.

However, on that commit, `NodeToNodeV_7` wasn't enabled in the `supportedNodeToNodeVersions` map for Cardano.

https://github.com/input-output-hk/ouroboros-network/blob/9b279c7548ee549e1ed755cd1acb69b6e69d0c7b/ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs#L315-L325

So I think it's still mutable. @coot @dcoutts confirm?

nfrisby replied:
@coot replied on Slack, confirming that the new KeepAlive is ready. So it's OK to also add Alonzo to NTN V_7 and then enable both Alonzo and KeepAlive2 at the same time 👍

nfrisby replied:
We also fixed-up our commit to set `latestReleasedNodeVersion` to NTN V6 and NTC V8 -- ie the ones that have been used by mainline releases. IE the predecessors of the versions that enable Alonzo (and as a coincidence enable KeepAlive2).

The resulting restriction of the node's default version negotation limit was the reason we added `latestReleasedNodeVersion` recently. And the corresponding new config file option `EnabledDevelopmentVersion` as of input-output-hk/cardano-node#2400 should let Benchmarking and Q&A etc _opt-in_ to negotiating the Alonzo-enabling versions.

EncodePanda reviewed (changes requested):
Small things need fixing. Those should be quick fixes though. Looking good, almost there :)

Also... no `Co-authored-by: EncodePanda <paul.szulc@gmail.com>`
![image](https://user-images.githubusercontent.com/6287558/118268427-aa112100-b4bd-11eb-871a-8e8219a0c01e.png)

EncodePanda commented:
It also seems we have to run it through stylish-haskell

nfrisby commented:
I opened Issue #3145 -- @Jasagredo and @michaelpj were discussing some of the options/observations on Slack just now.

We might need to explore some of those in order to get this building on Hydra.

nfrisby commented on ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Block.hs:
```diff
@@ -484,14 +551,29 @@ pattern QueryAnytimeMary
   -> CardanoQuery c result
 pattern QueryAnytimeMary q = QueryAnytime q (EraIndex (S (S (S (Z (K ()))))))
 
+-- | Query about the Mary era that can be answered anytime, i.e.,
```
Fixed!

nfrisby commented on ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Block.hs:
```diff
@@ -484,14 +551,29 @@ pattern QueryAnytimeMary
+-- For example, to ask for the start of the Alonzo era (whether the tip of the
+-- ledger is in the Byron, Shelley, ... era), use:
+--
+-- > QueryAnytimeMary EraStart
```
Fixed!

nfrisby commented on ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Block.hs:
```diff
@@ -484,14 +551,29 @@ pattern QueryAnytimeMary
+pattern QueryAnytimeAlonzo
+  :: QueryAnytime result
+  -> CardanoQuery c result
+pattern QueryAnytimeAlonzo q = QueryAnytime q (EraIndex (S (S (S (Z (K ()))))))
```
Fixed! Nice careful catch.

nfrisby replied:
I also pushed a new commit that refactors this module a bit to prevent this kind of mistake in the future. See the final commit on the PR.

nfrisby commented on ouroboros-consensus-cardano-test/test/Test/ThreadNet/MaryAlonzo.hs:
```diff
@@ -0,0 +1,384 @@
+{-# LANGUAGE DeriveGeneric        #-}
```
This whole module is a copy-paste. Again, we're kicking the can down the road: I'll be returning shortly to the ThreadNet rewrite work, and that will fully supplant this module and all those like it.

nfrisby commented:
I rebased to `git add` the new golden test files. We added new `Cardano` versions, so there are new goldens accordingly. However, as the following bash lines demonstrate, they're the same as they were for the previous version. That makes sense since our tech debt workaround was to not involve Alonzo in these Cardano tests yet. (We essentially only have the Mary-to-Alonzo hard fork ThreadNet test so far.)

```
$ git ls-files ouroboros-consensus-cardano-test/test/golden/CardanoNodeToNodeVersion5/ | awk '{ x = $0; y = x; gsub(/5/,"4",y); print "diff", $0, y; }' >/tmp/scr; . /tmp/scr; rm /tmp/scr
$ git ls-files ouroboros-consensus-cardano-test/test/golden/CardanoNodeToClientVersion7/ | awk '{ x = $0; y = x; gsub(/9/,"8",y); print "diff", $0, y; }' >/tmp/scr; . /tmp/scr; rm /tmp/scr
$ 
```

EncodePanda reviewed (approved):
LGTM

EncodePanda commented:
bors merge
